### ISSUE TITLE
Some SQL commands don't report a failure

### ISSUE THREAD

**Issue by visch** [dbt-msft/dbt-sqlserver#135]
EXEC('ALTER TABLE dbo.stg_things addd primary key(thing_id)')

This command will fail due to **addd**  being incorrect. This will just keep chugging along, is that just a side affect of EXEC that we have to live with? I'm going to attempt to not use EXEC, but I was warry when I read https://github.com/dbt-msft/dbt-sqlserver/issues/80 

Still new to DBT so I'm not certain how everything's supposed to work, but this almost certainly shouldn't happen. If a SQL command fails something should throw an error

**Comment by visch** [comment]
Changed the title as I just had this SQL run and fail, but dbt reported everything to be ok

```
   ALTER TABLE tablename
         ADD
         SysStartTime DATETIME2 GENERATED ALWAYS AS ROW START HIDDEN
            CONSTRAINT DF_SysStart DEFAULT SYSUTCDATETIME()
         ,SysEndTime DATETIME2 GENERATED ALWAYS AS ROW END HIDDEN
            CONSTRAINT DF_SysEnd DEFAULT CONVERT(DATETIME2, '9999-12-31 23:59:59.9999999'),
         PERIOD FOR SYSTEM_TIME (SysStartTime, SysEndTime);
         
         ALTER TABLE tablename 
         SET (SYSTEM_VERSIONING = ON (HISTORY_TABLE = tablename_history));
         ```


The failure was due to another table already having the DF_SysStart constraint setup. This can be replicated by setting up a table with this constraint before running the query.

**Comment by swanderz** [comment]
@visch very glad that you're putting dbt-sqlserver through it's paces here. my question is how are you running the code above? with a `post-hook`?

in [our related Slack thread](https://getdbt.slack.com/archives/CMRMDDQ9W/p1621019815068300), I mentioned how it might make sense to alter the create table macro to include an option to set `SYSTEM_VERSIONING`. Here's [the macro](https://github.com/dbt-msft/dbt-sqlserver/blob/83b8357519c585e4103481a5b987f0da6b376cce/dbt/include/sqlserver/macros/adapters.sql#L141-L165) you would need to change.

My problem is that dbt is all about using DML instead of DDL, which is what you'd need to create a Temporal Table...

**Comment by visch** [comment]
Yeah no problem glad to use this, I appreciate all the work everyone's put into this to allow me to get sql working very easily. Just a few hiccups :D 

I did the Primary Key, and Clustered Index in the create_table_as macro you mentioned under an if

{% if unique_key and not as_columnstore -%}

**Comment by visch** [comment]
I think these problems are coming from the EXEC statement

Did a little diving and found https://stackoverflow.com/a/55880440. Doing some kind of try catch when we need to do an EXEC might get us what we want here (proper failure) 

```
BEGIN TRY
    BEGIN TRY

        EXECUTE sp_executesql @stmt = N'SELECT 1/0;';
    END TRY
    BEGIN CATCH  

        DECLARE
             @ErrorNumber int
            ,@ErrorMessage nvarchar(2048)
            ,@ErrorSeverity int
            ,@ErrorState int
            ,@ErrorLine int;

        SELECT
             @ErrorNumber =ERROR_NUMBER()
            ,@ErrorMessage =ERROR_MESSAGE()
            ,@ErrorSeverity = ERROR_SEVERITY()
            ,@ErrorState =ERROR_STATE()
            ,@ErrorLine =ERROR_LINE();

        RAISERROR('Error %d caught in inner script at line %d: %s'
            ,@ErrorSeverity
            ,@ErrorState
            ,@ErrorNumber
            ,@ErrorLine
            ,@ErrorMessage);

    END CATCH; 
END TRY
BEGIN CATCH  

    THROW;

END CATCH; 
GO
```

**Comment by swanderz** [comment]
my question is still

> how are you running this code?

The whole design principle of dbt is that you never have to write Data Definition Language or Data Maniuplation Language, you just write select statements.

Maybe I'm missing something, but I'm struggling to understand how this a dbt problem, and not just an SQL Server problem. Are you struggling to modify tables that dbt created?

**Comment by visch** [comment]
I think I get where the misunderstanding is.

Let me restate this with a use case from dbt-sqlserver

https://github.com/dbt-msft/dbt-sqlserver/blob/master/dbt/include/sqlserver/macros/adapters.sql#L113 can fail due to Errors like source data not existing like this example https://github.com/dbt-msft/dbt-sqlserver/pull/126#issuecomment-848108760 

When this fails dbt doesn't report an error (`dbt run` reports a success on the model, when you'd expect a failure to be reporting).

**Comment by SportsTribution** [comment]
Hi,
to add a minimal example.
`SELECTx
    *
FROM {{source('source_data')}}`
leads in previous versions to:
`Database Error in model master_alarm (models\alarme\master_alarm.sql)
  ('42000', '[42000] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]Incorrect syntax near "SELECTx". (102) (SQLExecDirectW)')
  compiled SQL at target\run\mssql_test\models\alarme\master_alarm.sql

Done. PASS=1 WARN=0 ERROR=1 SKIP=0 TOTAL=2`
in 0.19.1 you simply get
`Completed successfully

Done. PASS=2 WARN=0 ERROR=0 SKIP=0 TOTAL=2`

All the best
Hannes

**Comment by swanderz** [comment]
ok @visch, now I understand. @SportsTribution can you tell me what version of dbt-sqlserver this worked? was it `0.19.0.1`?

@semcha @panasenco perhaps this has to do with how `EXEC` was introduced? I never thought to introduce a test that tries to make a broken model on purpose, but because we don't have one, everything worked...

https://github.com/dbt-msft/dbt-sqlserver/blob/679b3515e9f056879c6609c78bb62791716c450a/dbt/include/sqlserver/macros/adapters.sql#L156-L158

**Comment by SportsTribution** [comment]
Yes, it's `0.19.0.1`. I just tried `0.19.0.2` - works as well.
I think you're right. As I mentioned in https://github.com/dbt-msft/dbt-sqlserver/pull/126
I think it has something to do with pyodbc(?) not catching errors in the second command
https://github.com/flyway/flyway/issues/718 

So in 0.19.1 you have
USE [your_db];
execute(...)

And as there are two commands the second is somehow ignored in the error check

**Comment by visch** [comment]
Fixed with #140  !
